### DISCUSSION TITLE
Problems with turborepo caching

### DISCUSSION THREAD

**Discussion by Haberkamp** [vercel/turborepo#1755] [Help] [answered]
Hello, I'm trying to use turborepo with Vercel. Creating deployments on "development" branches work find but once I merge the PR on GitHub the deployment that gets created on the main branch fails. Maybe this has to do something with turborepo? The next.js project is not inside the root folder of the project but I changed the root folder for the Vercel project to the correct path.

<img width="1257" alt="Bildschirmfoto 2022-08-24 um 14 47 44" src="https://user-images.githubusercontent.com/35109813/186422867-533106e8-24d1-423e-bba8-c7c30dd51b0e.png">

**Comment by vitoorgomes** [comment]
Well, I was having this problem with cached builds all the time, but after better debugging I was able to fix it, I hope my solution works for you or anyone who faces the same issues.

### The Issue
First things first to be clear, this issue happens when Vercel tries to use the cached build to deploy the application. I know this is pretty straightforward, but hear me out. So, when Vercel will look for the `routes-manifest.json` it fails because it can't be found, again, pretty straightforward. While building locally you can easily find the `/path/your-app/.next/routes-manifest.json`, when on Vercel's platform while you are using turborepo, things don't work like this by default.

### The solution
For this part I just would like to point my configs, so it would be easier to compare:
1) My structure is something like this
```
root
|
├── apps
│   ├── app1
│   │   ├── ...
│   │   ├── .next
│   │   ├── .turbo
│   │   ├── package.json
│   ├── app2
│   │   ├── ...
│   │   ├── .next
│   │   ├── .turbo
│   │   ├── package.json
├── packages
│   ├── tsconfig
│   │   ├── ...
│   │   ├── package.json
│   ├── xxxx
│   │   ├── ...
│   │   ├── package.json
├── node_modules
├── package.json
├── yarn.lock
└── turbo.json
```

2) My turbo.json file is something like this
```json
{
  "$schema": "https://turbo.build/schema.json",
  "pipeline": {
    "build": {
      "outputs": [
        ".next/**"
      ],
      "inputs": [
        "src/**/*.{tsx,ts}"
      ]
    },
    "app1#build": {
      "dependsOn": [
        "^build"
      ],
      "outputs": [
        ".next/**"
      ],
      "env": [
       "MY_APP1_ENVS"
      ]
    },
    "app2#build": {
      "dependsOn": [
        "^build"
      ],
      "outputs": [
        ".next/**"
      ],
      "env": [
        "MY_APP2_ENVS"
      ]
    }
  }
}
```

3) My Vercel config
![image](https://user-images.githubusercontent.com/41302394/206881349-d4a6b749-dafc-4e0b-9981-822bdf0624ca.png)

With that in mind, I was having a lot of headache with this because I was not explicit setting the outputs paths on each of my app builds, that means that it just fallback to the default output folder
<img width="198" alt="image" src="https://user-images.githubusercontent.com/41302394/206881429-5357560b-65e2-4448-bbfe-612d7a401e7f.png">

Because I have each app with it's on build configs, just setting up the main `build.outputs: [".next/**"]` wasn't enough.  But after setting up on each `app#build` I got it working, because now it will output the build at the default place where Vercel is expecting (based on the Vercel config).

If you don't have each package having it's own build configs, adding the `outputs` on the main `build` command should do the trick.

![image](https://user-images.githubusercontent.com/41302394/206881518-6e6014e7-f263-4360-bc13-b0f954212ffe.png)

Just one last thing, I was able to track this by using the `--dry=json` flag when executing the `turbo run build`, this prevents the command to execute, but shows the log of what it's happening [docs](https://turbo.build/repo/docs/reference/command-line-reference#--dry----dry-run)

**Comment by anthonyshew** [comment, marked as answer]
Check out this documentation for how to deploy a Turborepo on Vercel: https://vercel.com/docs/monorepos/turborepo
